waterdog/src/main/java/me/lucko/luckperms/waterdog/LPWaterdogPlugin.java

```diff
@@ -0,0 +1,257 @@
+import java.util.Set;
+import java.util.stream.Stream;
+
+/**
+ * LuckPerms implementation for the Waterdog API.
+ */
+public class LPWaterdogPlugin extends AbstractLuckPermsPlugin {
+    private final LPWaterdogBootstrap bootstrap;
+
+    private WaterdogSenderFactory senderFactory;
+    private WaterdogConnectionListener connectionListener;
+    private CommandManager commandManager;
+    private StandardUserManager userManager;
+    private StandardGroupManager groupManager;
+    private StandardTrackManager trackManager;
+    private WaterdogContextManager contextManager;
+
+    public LPWaterdogPlugin(LPWaterdogBootstrap bootstrap) {
+        this.bootstrap = bootstrap;
+    }
+
+    @Override
+    public LPWaterdogBootstrap getBootstrap() {
+        return this.bootstrap;
+    }
+
+    public Plugin getLoader() {
+        return this.bootstrap.getLoader();
+    }
+
+    @Override
+    protected void setupSenderFactory() {
+        this.senderFactory = new WaterdogSenderFactory(this);
+    }
+
+    @Override
+    protected ConfigurationAdapter provideConfigurationAdapter() {
+        return new WaterdogConfigAdapter(this, resolveConfig("config.yml").toFile());
+    }
+
+    @Override
+    protected void registerPlatformListeners() {
+        this.connectionListener = new WaterdogConnectionListener(this);
+
+        WaterdogPermissionCheckListener permissionCheckListener = new WaterdogPermissionCheckListener(this);
+
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                PlayerPermissionCheckEvent.class,
+                permissionCheckListener::onPlayerPermissionCheck,
+                EventPriority.HIGH
+        );
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                PlayerPermissionCheckEvent.class,
+                permissionCheckListener::onOtherPermissionCheck,
+                EventPriority.HIGHEST
+        );
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                TristateCheckEvent.class,
+                permissionCheckListener::onPlayerTristateCheck,
+                EventPriority.HIGH
+        );
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                TristateCheckEvent.class,
+                permissionCheckListener::onOtherTristateCheck,
+                EventPriority.HIGHEST
+        );
+
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                PlayerDisconnectEvent.class,
+                this.connectionListener::onPlayerQuit,
+                EventPriority.LOW
+        );
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                PlayerLoginEvent.class,
+                this.connectionListener::onPlayerLogin,
+                EventPriority.LOWEST
+        );
+        this.bootstrap.getProxy().getEventManager().subscribe(
+                PlayerLoginEvent.class,
+                this.connectionListener::onPlayerPostLogin,
+                EventPriority.LOW
+        );
```

Comment:
Instead of having these `subscribe` calls in the plugin class, I think it would be better if it was done in the listeners themselves.

My preference would be to create a method `public void register(EventManager eventManager) {...}` in each of the listener classes to do the subscribing.

After that's done, we can make the `onBlahWhateter` methods private.